File: src/components/reorder.rs
//! Lists whose rows can be dragged, or moved with the keyboard, into a new
//! order.
use super::gesture::{GestureScript, use_gesture};
use crate::state::{provide_live_context, use_element_id, use_live_context};
use crate::theme::{merge_classes, use_strings};
use dioxus::prelude::*;
use dioxus_icons::lucide::GripVertical;
use std::collections::HashMap;
use std::rc::Rc;

/// Which edge of a row holds a [`ReorderHandle`].
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Default)]
pub enum ReorderHandlePosition {
    /// Leading edge.
    #[default]
    Start,
    /// Trailing edge, matching Ionic's default.
    End,
}

/// How the items of a [`ReorderList`] are laid out, which decides where a
/// dragged item lands and how the others make room.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Default)]
pub enum ReorderLayout {
    /// One item per row, top to bottom. A dragged row takes the place of
    /// each row whose middle it passes, and the rows between move up or down
    /// by its height, so rows of any height work.
    #[default]
    Rows,
    /// Items that wrap across rows in reading order, such as the cells of a
    /// [`Grid`](crate::Grid). A dragged item lands on the cell whose middle
    /// is nearest the pointer, and the cells between each step into the
    /// neighbouring slot, wrapping between rows. Suits cells of one size.
    Grid,
}

impl ReorderLayout {
    fn as_str(self) -> &'static str {
        match self {
            ReorderLayout::Rows => "rows",
            ReorderLayout::Grid => "grid",
        }
    }
}

/// Follows a dragged item in the webview, moving the others aside, and
/// reports where it was dropped. See [`super::gesture`] for why.
const REORDER_SCRIPT: GestureScript = GestureScript {
    name: "g3-ui.reorder",
    source: include_str!("reorder.js"),
};

#[derive(Clone, Copy, PartialEq)]
struct ReorderContext {
    /// Every item's element, by position, so the keyboard knows how many
    /// places there are.
    rows: Signal<HashMap<usize, Rc<MountedData>>>,
    announcement: Signal<String>,
    onreorder: EventHandler<(usize, usize)>,
    disabled: Signal<bool>,
    layout: Signal<ReorderLayout>,
}

/// A list whose rows can be put in a new order. Like Ionic's
/// `ion-reorder-group`.
///
/// Wrap each row in a [`ReorderItem`] with its position, and put a
/// [`ReorderHandle`] in the row. Dragging the handle moves the row, and the
/// others make room; the arrow keys move it one place while the handle has
/// focus. Either way `onreorder` is called with the row's old and new
/// positions, and the list is reordered by the caller, which owns the data:
///
/// ```
/// # use dioxus::prelude::*;
/// # use g3_ui::prelude::*;
/// # fn demo() -> Element {
/// let mut holes = use_signal(|| vec!["Front nine", "Back nine", "Par threes"]);
/// rsx! {
///     ReorderList {
///         onreorder: move |(from, to): (usize, usize)| {
///             holes.with_mut(|holes| {
///                 let hole = holes.remove(from);
///                 holes.insert(to, hole);
///             });
///         },
///         List { variant: ListVariant::Raised,
///             for (index, hole) in holes().into_iter().enumerate() {
///                 ReorderItem { key: "{hole}", index,
///                     Item {
///                         label: hole,
///                         start: rsx! { ReorderHandle { label: format!("Move {hole}") } },
///                     }
///                 }
///             }
///         }
///     }
/// }
/// # }
/// ```
///
/// With [`ReorderLayout::Grid`] the items can wrap across rows, as the cells
/// of a [`Grid`](crate::Grid) do, and the left and right arrows move a cell
/// too. The list does not lay its items out itself, so any grid works:
///
/// ```
/// # use dioxus::prelude::*;
/// # use g3_ui::prelude::*;
/// # fn demo() -> Element {
/// let mut posters = use_signal(|| vec!["Dune", "Heat", "Alien", "Up"]);
/// rsx! {
///     ReorderList {
///         layout: ReorderLayout::Grid,
///         onreorder: move |(from, to): (usize, usize)| {
///             posters.with_mut(|posters| {
///                 let poster = posters.remove(from);
///                 posters.insert(to, poster);
///             });
///         },
///         Grid { columns: GridColumns::Count(2),
///             for (index, title) in posters().into_iter().enumerate() {
///                 ReorderItem { key: "{title}", index,
///                     Card { title: title.to_string(),
///                         ReorderHandle { label: format!("Move {title}") }
///                     }
///                 }
///             }
///         }
///     }
/// }
/// # }
/// ```
#[component]
pub fn ReorderList(
    /// Called with `(from, to)` when a row is moved.
    onreorder: EventHandler<(usize, usize)>,
    /// How the items are laid out. Defaults to [`ReorderLayout::Rows`].
    layout: Option<ReorderLayout>,
    /// Turn reordering off, leaving the rows where they are.
    disabled: Option<bool>,
    /// Extra classes for the list wrapper.
    class: Option<String>,
    /// The list, whose rows are [`ReorderItem`]s.
    children: Element,
) -> Element {
    let id = use_element_id("reorder", None);
    let rows = use_signal(HashMap::<usize, Rc<MountedData>>::new);
    let disabled = crate::state::use_synced_signal(disabled.unwrap_or(false));
    let announcement = use_signal(String::new);
    let layout = crate::state::use_synced_signal(layout.unwrap_or_default());
    let context = ReorderContext {
        rows,
        announcement,
        onreorder,
        disabled,
        layout,
    };
    provide_live_context(context);

    // The script moves the rows while they are dragged; a drop that changes
    // the order arrives here once.
    let start_script = use_gesture(REORDER_SCRIPT, id.clone(), move |gesture| {
        if gesture.kind == "reorder" && !disabled() {
            let (from, to) = (gesture.value(0) as usize, gesture.value(1) as usize);
            if from != to {
                onreorder((from, to));
            }
        }
    });

    rsx! {
        div {
            id,
            class: merge_classes("g3-reorder", class.as_deref()),
            // Read by the script at each press.
            "data-layout": layout().as_str(),
            onmounted: move |_| start_script.call(()),
            {children}
            span { class: "g3-sr-only", role: "status", aria_live: "polite",
                "{announcement}"
            }
        }
    }
}

/// One row of a [`ReorderList`], at position `index`.
#[component]
pub fn ReorderItem(
    /// The row's position in the list.
    index: usize,
    /// Extra classes for the row wrapper.
    class: Option<String>,
    children: Element,
) -> Element {
    let context = use_live_context::<ReorderContext>();
    let mut rows = context.rows;
    // Rows are keyed, so a reorder moves them without remounting. Their
    // position is a signal, so the handle inside follows it, and each row
    // re-registers its element under its new position.
    let position = crate::state::use_synced_signal(index);
    use_context_provider(|| ReorderPosition(position));
    let mut mounted = use_signal(|| None::<Rc<MountedData>>);
    use_effect(move || {
        let index = position();
        if let Some(element) = mounted() {
            rows.write().insert(index, element);
        }
    });
    use_drop(move || {
        let index = *position.peek();
        let ours = match (rows.peek().get(&index), mounted.peek().as_ref()) {
            (Some(registered), Some(element)) => Rc::ptr_eq(registered, element),
            _ => false,
        };
        if ours {
            rows.write().remove(&index);
        }
    });
    rsx! {
        div {
            class: merge_classes("g3-reorder-item", class.as_deref()),
            // The script orders the rows by this, and waits for it to change
            // before letting go of a dropped row.
            "data-index": "{index}",
            onmounted: move |event: MountedEvent| mounted.set(Some(event.data())),
            {children}
        }
    }
}

/// The position a [`ReorderItem`] gives its handle.
#[derive(Clone, Copy)]
struct ReorderPosition(Signal<usize>);

/// The grip that moves a row of a [`ReorderList`]. Drag it, or focus it and
/// press the up and down arrows; in a [`ReorderLayout::Grid`], left and
/// right move it back and on too.
#[component]
pub fn ReorderHandle(
    /// Accessible name, naming the row, such as "Move Dune". Defaults to
    /// [`Strings::reorder`](crate::Strings::reorder).
    label: Option<String>,
    /// Edge spacing for the handle. Place the handle in the matching `Item`
    /// slot. Defaults to [`ReorderHandlePosition::Start`] for compatibility
    /// with the original handle API.
    position: Option<ReorderHandlePosition>,
    /// Extra classes for the handle.
    class: Option<String>,
) -> Element {
    let context = use_live_context::<ReorderContext>();
    let ReorderPosition(row_position) = use_context::<ReorderPosition>();
    let index = row_position();
    let mut announcement = context.announcement;
    let label = label.unwrap_or_else(|| use_strings().reorder);
    let disabled = (context.disabled)();
    let position = position.unwrap_or_default();
    let rows = context.rows;

    let count = rows.read().len();
    let grid = (context.layout)() == ReorderLayout::Grid;
    let key = move |event: KeyboardEvent| {
        if disabled {
            return;
        }
        // In a grid the item moves along the reading order, so left and
        // right move it too.
        let back = matches!(event.key(), Key::ArrowUp) || (grid && event.key() == Key::ArrowLeft);
        let on = matches!(event.key(), Key::ArrowDown) || (grid && event.key() == Key::ArrowRight);
        let to = match (back, on) {
            (true, _) if index > 0 => index - 1,
            (_, true) if index + 1 < count => index + 1,
            _ => return,
        };
        event.prevent_default();
        (context.onreorder)((index, to));
        // The handle keeps focus and its name; only the new place is news.
        announcement.set(format!("{} / {count}", to + 1));
    };

    rsx! {
        button {
            r#type: "button",
            class: merge_classes("g3-reorder-handle", class.as_deref()),
            "data-position": match position {
                ReorderHandlePosition::Start => "start",
                ReorderHandlePosition::End => "end",
            },
            aria_label: label.clone(),
            // Also how the script knows not to start a drag.
            aria_disabled: disabled.then_some("true"),
            onkeydown: key,
            // A press on the handle is for dragging, not for the row.
            onclick: move |event| event.stop_propagation(),
            GripVertical { size: 20 }
        }
    }
}

#[cfg(feature = "playground")]
#[component]
fn ReorderPlaygroundDemo() -> Element {
    let mut rows = use_signal(|| {
        vec![
            "Front nine",
            "Back nine",
            "Par threes",
            "Longest drive",
            "Closest to the pin",
            "Best round",
        ]
    });
    let position = use_signal(|| ReorderHandlePosition::End);
    let layout = use_signal(ReorderLayout::default);
    rsx! {
        crate::PlaygroundDemoFrame { center: false,
            controls: rsx! {
                crate::SegmentGroup { value: layout, aria_label: "Layout",
                    crate::SegmentButton { value: ReorderLayout::Rows, "Rows" }
                    crate::SegmentButton { value: ReorderLayout::Grid, "Grid" }
                }
                if layout() == ReorderLayout::Rows {
                    crate::SegmentGroup { value: position, aria_label: "Handle edge",
                        crate::SegmentButton { value: ReorderHandlePosition::Start, "Start" }
                        crate::SegmentButton { value: ReorderHandlePosition::End, "End" }
                    }
                }
            },
            ReorderList {
                layout: layout(),
                onreorder: move |(from, to): (usize, usize)| {
                    rows.with_mut(|rows| {
                        let row = rows.remove(from);
                        rows.insert(to, row);
                    });
                },
                if layout() == ReorderLayout::Grid {
                    crate::Grid { columns: crate::GridColumns::Count(3), gap: crate::Space::Sm,
                        for (index, row) in rows().into_iter().enumerate() {
                            ReorderItem { key: "{row}", index,
                                crate::Card {
                                    title: row.to_string(),
                                    subtitle: format!("Position {}", index + 1),
                                    ReorderHandle { label: format!("Move {row}") }
                                }
                            }
                        }
                    }
                } else {
                    crate::List { variant: crate::ListVariant::Raised,
                        for (index, row) in rows().into_iter().enumerate() {
                            ReorderItem { key: "{row}", index,
                                crate::Item {
                                    label: row,
                                    description: format!("Position {}", index + 1),
                                    start: (position() == ReorderHandlePosition::Start).then(|| rsx! {
                                        ReorderHandle { label: format!("Move {row}"), position: ReorderHandlePosition::Start }
                                    }),
                                    end: (position() == ReorderHandlePosition::End).then(|| rsx! {
                                        ReorderHandle { label: format!("Move {row}"), position: ReorderHandlePosition::End }
                                    }),
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

crate::g3_playground! {
    name: "Reorder",
    description: "Rows or grid cells moved by dragging a handle, or with the arrow keys.",
    components: ["ReorderList", "ReorderItem", "ReorderHandle"],
    demo: ReorderPlaygroundDemo,
    source: "src/components/reorder.rs",
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn script_stays_alive_and_never_waits_on_rust() {
        super::super::gesture::assert_gesture_script(REORDER_SCRIPT);
    }
}

File: src/components/reorder.js
// Drives every ReorderList on the page from inside the webview, so a dragged
// item follows the pointer without waiting on Rust (see gesture.rs). Rust
// hears one thing per drag: the item's old and new positions, if it moved.
// A list's `data-layout` says whether its items are rows or grid cells.
const key = Symbol.for("g3-ui.reorder");
window[key]?.dispose?.();

// Longest a dropped row waits for the caller to reorder the list before it
// gives up and slides back.
const SETTLE_MS = 400;

// How long a dropped item takes to glide from the pointer into its slot;
// matches `--g3-reorder-drop` in the stylesheet.
const DROP_MS = 200;

let active = null;

// The list's own rows, in order, skipping any in a list nested inside it.
const rowsOf = (list) => [...list.querySelectorAll(".g3-reorder-item")]
    .filter((row) => row.closest(".g3-reorder") === list)
    .sort((a, b) => Number(a.dataset.index) - Number(b.dataset.index));

// Where an item dragged from `from` would land, given the pointer and the
// middle of every item before the drag began.
//
// Rows: past the middle of each row the pointer has crossed, so rows of any
// height work. Grid: the cell whose middle is nearest, wherever it wraps.
const targetIndex = (drag, x, y) => {
    if (drag.grid) {
        let target = drag.from;
        let nearest = Infinity;
        drag.centers.forEach(([cx, cy], index) => {
            const distance = (cx - x) ** 2 + (cy - y) ** 2;
            if (distance < nearest) {
                nearest = distance;
                target = index;
            }
        });
        return target;
    }
    let target = drag.from;
    drag.centers.forEach(([, middle], index) => {
        if (index < drag.from && y < middle) target = Math.min(target, index);
        if (index > drag.from && y > middle) target = Math.max(target, index);
    });
    return target;
};

// How far an item that is not being dragged moves aside for the one that
// is: toward the place the dragged item left, when the drag passes over it.
// A row moves by the dragged row's height; a cell steps into the next slot,
// which may be at the other end of the row above or below.
const shift = (index, drag) => {
    let slot;
    if (drag.from < drag.target && index > drag.from && index <= drag.target) slot = index - 1;
    else if (drag.target < drag.from && index >= drag.target && index < drag.from) slot = index + 1;
    else return [0, 0];
    if (!drag.grid) return [0, slot < index ? -drag.height : drag.height];
    const [ox, oy] = drag.centers[index];
    const [sx, sy] = drag.centers[slot];
    return [sx - ox, sy - oy];
};

const paint = (drag) => {
    drag.rows.forEach((row, index) => {
        const [x, y] = index === drag.from ? [drag.grid ? drag.dx : 0, drag.dy] : shift(index, drag);
        if (x || y) row.style.transform = `translate3d(${x}px, ${y}px, 0)`;
        else row.style.removeProperty("transform");
    });
};

const clear = (drag) => {
    for (const row of drag.rows) {
        row.style.removeProperty("transform");
        delete row.dataset.dragging;
        delete row.dataset.settling;
    }
    delete drag.list.dataset.dragging;
};

// Where the dragged item's slot is, measured from where it started.
const slotOffset = (drag) => {
    if (drag.grid) {
        const [ox, oy] = drag.centers[drag.from];
        const [tx, ty] = drag.centers[drag.target];
        return [tx - ox, ty - oy];
    }
    const from = drag.rects[drag.from];
    const to = drag.rects[drag.target];
    return [0, drag.target > drag.from ? to.bottom - from.bottom : to.top - from.top];
};

// Glides the dragged item from the pointer to `[x, y]`, then calls `then`.
// Without it, a dropped item jumped from under the pointer into its slot.
const glide = (drag, [x, y], then) => {
    const row = drag.rows[drag.from];
    const still = (!drag.grid || drag.dx === x) && drag.dy === y;
    if (still || matchMedia("(prefers-reduced-motion: reduce)").matches) return then();
    let finished = false;
    const finish = () => {
        if (finished) return;
        finished = true;
        row.removeEventListener("transitionend", onEnd);
        clearTimeout(timer);
        then();
    };
    const onEnd = (event) => {
        if (event.target === row && event.propertyName === "transform") finish();
    };
    row.addEventListener("transitionend", onEnd);
    // In case the transition never runs, as when the row is hidden.
    const timer = setTimeout(finish, DROP_MS + 80);
    row.dataset.settling = "true";
    if (x || y) row.style.transform = `translate3d(${x}px, ${y}px, 0)`;
    else row.style.removeProperty("transform");
};

// Capturing, and stopped here, so the press is the handle's alone: neither a
// swipe row around it nor anything in Rust hears it.
const onPointerDown = (event) => {
    const handle = event.target instanceof Element && event.target.closest(".g3-reorder-handle");
    const list = handle?.closest(".g3-reorder");
    const row = handle?.closest(".g3-reorder-item");
    // A list still settling its last drop takes no new drag until it has.
    if (!list || !row || active || list.dataset.dragging === "true"
        || handle.getAttribute("aria-disabled") === "true") return;
    event.preventDefault();
    event.stopPropagation();
    const rows = rowsOf(list);
    const from = rows.indexOf(row);
    if (from < 0) return;
    // Measured as the drag begins, so items of any size make room for each
    // other correctly.
    const rects = rows.map((each) => each.getBoundingClientRect());
    active = {
        list,
        rows,
        grid: list.dataset.layout === "grid",
        pointerId: event.pointerId,
        from,
        target: from,
        startX: event.clientX,
        startY: event.clientY,
        dx: 0,
        dy: 0,
        height: rects[from].height,
        rects,
        centers: rects.map((rect) => [rect.left + rect.width / 2, rect.top + rect.height / 2]),
    };
    try { handle.setPointerCapture(event.pointerId); } catch (error) {}
    list.dataset.dragging = "true";
    row.dataset.dragging = "true";
};

const onPointerMove = (event) => {
    const drag = active;
    if (!drag || event.pointerId !== drag.pointerId) return;
    drag.dx = event.clientX - drag.startX;
    drag.dy = event.clientY - drag.startY;
    drag.target = targetIndex(drag, event.clientX, event.clientY);
    paint(drag);
};

const onPointerUp = (event) => {
    const drag = active;
    if (!drag || event.pointerId !== drag.pointerId) return;
    active = null;
    if (drag.target === drag.from) return putBack(drag);
    glide(drag, slotOffset(drag), () => {
        // Hold every row where the drop left it until the caller's reorder
        // reaches the page, then let go in the same frame, so the dropped
        // row never flashes back to where it started.
        let settled = false;
        const settle = () => {
            if (settled) return;
            settled = true;
            observer.disconnect();
            clearTimeout(timer);
            clear(drag);
        };
        const observer = new MutationObserver(settle);
        observer.observe(drag.list, {
            subtree: true,
            childList: true,
            attributes: true,
            attributeFilter: ["data-index"],
        });
        const timer = setTimeout(settle, SETTLE_MS);
        dioxus.send([drag.list.id, "reorder", [drag.from, drag.target]]);
    });
};

// Slides everything back to where the drag began.
const putBack = (drag) => {
    drag.target = drag.from;
    drag.rows.forEach((row, index) => {
        if (index !== drag.from) row.style.removeProperty("transform");
    });
    glide(drag, [0, 0], () => clear(drag));
};

const onPointerCancel = (event) => {
    const drag = active;
    if (!drag || event.pointerId !== drag.pointerId) return;
    active = null;
    putBack(drag);
};

document.addEventListener("pointerdown", onPointerDown, true);
document.addEventListener("pointermove", onPointerMove);
document.addEventListener("pointerup", onPointerUp);
document.addEventListener("pointercancel", onPointerCancel);

let released;
const done = new Promise((resolve) => {
    released = resolve;
});
window[key] = {
    dispose() {
        document.removeEventListener("pointerdown", onPointerDown, true);
        document.removeEventListener("pointermove", onPointerMove);
        document.removeEventListener("pointerup", onPointerUp);
        document.removeEventListener("pointercancel", onPointerCancel);
        released();
    },
};
// Native renderers close an eval's channel as soon as its script returns.
await done;
